src/main/java/eu/europeana/metis/sandbox/service/problempatterns/PatternAnalysisServiceImpl.java

```diff
@@ -155,10 +151,40 @@ public void generateRecordPatternAnalysis(String datasetId, Step executionStep,
 
   @Override
   @Transactional
-  public void finalizeDatasetPatternAnalysis(String datasetId, Step executionStep, LocalDateTime executionTimestamp) {
+  public void finalizeDatasetPatternAnalysis(ExecutionPoint executionPoint) {
     //This is currently meant to be implemented for the P1 with the titles which will come later on.
   }
 
+  private ArrayList<ProblemPattern> constructProblemPatterns(ExecutionPoint executionPoint) {
+    final ArrayList<ProblemPattern> problemPatterns = new ArrayList<>();
+    for (DatasetProblemPattern datasetProblemPattern : executionPoint.getDatasetProblemPatterns()) {
+
+      final ArrayList<RecordAnalysis> recordAnalyses = getRecordAnalyses(executionPoint, datasetProblemPattern);
+      if (CollectionUtils.isNotEmpty(recordAnalyses)) {
+        problemPatterns.add(new ProblemPattern(
+            ProblemPatternDescription.fromName(datasetProblemPattern.getDatasetProblemPatternId().getPatternId()),
+            datasetProblemPattern.getRecordOccurrences(), recordAnalyses));
+      }
+    }
+    return problemPatterns;
+  }
+
+  private ArrayList<RecordAnalysis> getRecordAnalyses(ExecutionPoint executionPoint,
+      DatasetProblemPattern datasetProblemPattern) {
+    final ArrayList<RecordAnalysis> recordAnalyses = new ArrayList<>();
+    executionPoint.getRecordProblemPatterns().forEach(recordProblemPattern -> {
+      final ArrayList<ProblemOccurrence> problemOccurrences = new ArrayList<>();
+      for (RecordProblemPatternOccurrence recordProblemPatternOccurrence : recordProblemPattern.getRecordProblemPatternOccurences()) {
+        problemOccurrences.add(new ProblemOccurrence(recordProblemPatternOccurrence.getMessageReport()));
+      }
+      //Select only the relevant ones
+      if (datasetProblemPattern.getDatasetProblemPatternId().getPatternId().equals(recordProblemPattern.getPatternId())) {
```

Comment:
Is it possible to rename this one to ```DatasetProblemPatternEmbedded``` or ```DatasetProblemPatternKey``` or ```DatasetProblemComposite``` or ```DatasetProblemCompositeKey``` it was confusing reading to get the ```Id``` of the ```Id```

Reply:
I've updated it a bit, hope it's clear now.

Reply:
Well those are two different ids.
The DatasetProblemPatternId is the composite id used for the DatasetProblemPattern, therefore the equivalent name.
And the patternId is the id of the pattern per case.

Reply:
I've renamed to DatasetProblemCompositeKey